geocruncher/voxel_computation.py

```diff
@@ -0,0 +1,62 @@
+import sys
+import os
+import numpy as np
+from gmlib.GeologicalModel3D import GeologicalModel
+from gmlib.architecture import from_GeoModeller, Evaluator
+import pyvista as pv
+
+
+def _compute_voxels(res, box, model, meshes_files, out_file):
+    # args: [1] resolution [2] box 3d of the projet [3] model geologic of the project [4] list of filename of meshes [5] output file
+    ny, nx, nz = res
+
+    # we use numpy meshgrid to produce a regular grid
+    # the output is a list containing a 3D array for each coordinate
+    # if we want an evaluation on the center of the voxels
+    # we would have to compute voxel dimensions
+    dx = (box.xmax - box.xmin) / nx
+    dy = (box.ymax - box.ymin) / ny
+    dz = (box.zmax - box.zmin) / nz
+
+    x, y, z = np.meshgrid(
+        np.arange(box.xmin + 0.5 * dx, box.xmax, dx),
+        np.arange(box.ymin + 0.5 * dy, box.ymax, dy),
+        np.arange(box.zmin + 0.5 * dz, box.zmax, dz),
+    )
+
+    # we transform the previous data into an array of 3D points
+    # this could be vertices from an unstructured mesh or any other data
+    xyz = np.stack((x, y, z), axis=-1)
+    xyz.shape = (-1, 3)
+
+    gwb_tags = [0] * xyz.shape[0]
+    for mesh_file in meshes_files:
+        gwb_id = int(mesh_file.split("_")[1])
+        mesh = pv.read(mesh_file)
+        mesh = mesh.extract_geometry()
+        points = pv.PolyData(xyz)
+
+        insidePoints = points.select_enclosed_points(mesh, tolerance=0.00001)
+        gwb_tags = [max(newId, _id) for newId, _id in zip(insidePoints["SelectedPoints"] * gwb_id, gwb_tags)]
+
+    ranks = list(map(lambda point:  model.rank(point, True), xyz))
+
+    # More performant version but there is a bug with topography
```

Comment:
Maybe you could try the new gmlib version I have just updated the geocruncher PR.
Ps: I haven't tested it yet.